### PR TITLE
clean up engine.cpp thread state

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63117
* #63116
* __->__ #63115
* #63860

This actually changes:
- callbacks now run with proper grad mode even in worker threads
- graphtask's Future callbacks now run with proper TLS when erroring
  out from a worker thread

Differential Revision: [D30388100](https://our.internmc.facebook.com/intern/diff/D30388100)

### CODE REVIEW

File: torch/csrc/autograd/engine.cpp [hunk 1]
```diff
@@ -407,7 +407,12 @@ auto Engine::thread_main(const std::shared_ptr<GraphTask>& graph_task) -> void {
       }
 
       if (task.fn_ && !local_graph_task->has_error_.load()) {
+        // Set the ThreadLocalState before calling the function.
+        // NB: The ThreadLocalStateGuard doesn't set the grad_mode because GraphTask
+        // always saves ThreadLocalState without grad_mode.
```
Comment (soulitzer) [inline torch/csrc/autograd/engine.cpp:412]: Noob question: what was the original reason for this comment - why can't we just have GraphTask save ThreadLocalState with grad_mode? Can we just have thread_locals_ in GraphTask keep_grad_mode=true now that we have both guards in the same place (maybe there are some instances that I am missing)?

**Review by soulitzer — APPROVED** [review]

File: torch/csrc/autograd/engine.cpp (hunk 1 above)
Comment (albanD) [reply torch/csrc/autograd/engine.cpp:412]: The gradmode is controlled by the `create_graph` flag passed to backward/grad. Not the current state of the thread.

**Comment by albanD** [comment]
For some reason ghstack didn't like my rebasing work and pushed a brand new stack (maybe I shouldn't have upgraded ghstack?)
In any case, the new version of this PR is here: https://github.com/pytorch/pytorch/pull/63318

Sorry for the noise.

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388100).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388100).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388100).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30388100).

**Pull request closed without merging**